docs/large-tutorials/pyro.md

```diff
@@ -0,0 +1,1043 @@
+```
+
+```{code-cell} ipython3
+model.to(device)
+
+# define SVI guide
+guide = AutoMultivariateNormal(model, init_loc_fn=init_to_mean)
+
+adam = pyro.optim.Adam({"lr": 0.02})
+svi = SVI(model, guide, adam, loss=Trace_ELBO())
+
+num_iterations = 15000
+pyro.clear_param_store()
+loss_tracker = np.empty(num_iterations)
+
+t0 = time.time()
+for j in range(num_iterations):
+    # calculate the loss and take a gradient step
+    loss_tracker[j] = svi.step(predictive=False)
+    
+print("Optimization took {:}s".format(time.time() - t0))
+
+# write loss to file 
+table = Table()
+table["loss"] = np.array(loss_tracker)
+ascii.write(table, "loss.csv", overwrite=True)
+```
+
+```{code-cell} ipython3
+table = ascii.read("loss.csv")
+# subtract the minimum value 
+loss = table["loss"]
+loss -= np.min(loss)
+
+# plot loss
+fig, ax = plt.subplots(nrows=1)
+ax.semilogy(loss)
+ax.set_xlabel("iteration")
+ax.set_ylabel("loss");
+```
+
+### Visualization of samples
+
+We'll follow a similar procedure as with the `AutoNormal` guide.
+
+```{code-cell} ipython3
+samples = Predictive(model, guide=guide, return_sites=['disk.incl', 'disk.Omega', 'disk.x_centroid', 'disk.y_centroid', 'disk.log_A_0', 'disk.log_sigma_0', 'disk.log_ring_amplitudes', 'disk.ring_means', 'disk.log_ring_sigmas'], num_samples=2000)(True)
+for k, v in samples.items():
+    print(f"{k}: {v.shape}")
+```
+
+```{code-cell} ipython3
+# extract samples from the Pyro Predictive object and convert units for convenience
+dict_samples = {k: np.expand_dims(v.detach().numpy(), 0) for k, v in samples.items()}
+# convert from radians to degrees
+for key in ["disk.incl", "disk.Omega"]:
+    dict_samples[key] /= deg
+    
+# convert to actual value
+for key in ["disk.log_A_0", "disk.log_sigma_0", "disk.log_ring_amplitudes", "disk.log_ring_sigmas"]:
+    new_key = key.replace("log_", "")
+    dict_samples[new_key] = 10**dict_samples.pop(key)    
+    
+dataset = az.convert_to_inference_data(dict_samples)
+```
+
+Because it is hard to visualize the posteriors for all 27 parameters in a single plot, we will plot pairwise plots of a subset of the variables at a time.
+
+```{code-cell} ipython3
+az.plot_pair(dataset, var_names=["disk.ring_means"]);
+```
+
+```{code-cell} ipython3
+az.plot_pair(dataset, var_names=["disk.ring_sigmas"]);
+```
+
+```{code-cell} ipython3
+az.plot_pair(dataset, var_names=["disk.ring_amplitudes"]);
+```
+
+With the more flexible guide, the correlations between parameters are more accurately captured. Now let's see what the model and residuals look like for this optimized posterior distribution.
+
+```{code-cell} ipython3
+samples = Predictive(model, guide=guide, return_sites=['vis_real', 'vis_imag', 'sky_cube'], num_samples=1)(predictive=True)
+fig = compare_dirty_model_resid(samples["vis_real"][0], samples["vis_imag"][0], samples["sky_cube"][0]);
+```
+
+It's hard to tell much of a difference with the model and residual images.
```

Comment:
I think the difference in the residual image is dramatic relative to the AutoNormal result; the peak residual amplitude has decreased by a factor of ~2, and the structure in the residual image is reduced quite a bit. Again I'd say that a different colorscale would make this difference more stark.

Reply:
We're drawing random samples from the posterior, so I think, when averaged over multiple samples, the residual change is not that significant (when viewed with the changed colorbar). I could make a side-by-side movie comparing draws from the two, but this is a bit more work.